src/backend/env/invite.rs

```diff
@@ -0,0 +1,138 @@
+use super::{user::UserId, Credits, Principal, RealmId, State, User};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeSet;
+
+#[derive(Default, Serialize, Deserialize, Clone)]
+pub struct Invite {
+    pub credits: Credits,
+    pub credits_per_user: Credits,
+    pub joined_user_ids: BTreeSet<UserId>,
+    pub realm_id: Option<RealmId>,
+    /// Owner id
+    pub user_id: UserId,
+}
+
+impl Invite {
+    pub fn new(
+        credits: Credits,
+        credits_per_user: Credits,
+        realm_id: Option<RealmId>,
+        user_id: UserId,
+    ) -> Self {
+        // Convert empty realm_id to None
+        let converted_relm_id: Option<RealmId> = realm_id.filter(|id| !id.is_empty());
+
+        Self {
+            credits,
+            credits_per_user,
+            joined_user_ids: BTreeSet::new(),
+            realm_id: converted_relm_id,
+            user_id,
+        }
+    }
+
+    pub fn consume(&mut self, joined_user_id: UserId) -> Result<(), String> {
+        let new_credits = self
+            .credits
+            .checked_sub(self.credits_per_user)
+            .ok_or("Invite credits too low")?;
+        self.credits = new_credits;
+
+        if self.joined_user_ids.contains(&joined_user_id) {
+            return Err("User already credited".into());
+        }
+
+        self.joined_user_ids.insert(joined_user_id);
+
+        Ok(())
+    }
+
+    pub fn update(
+        &mut self,
+        credits: Option<Credits>,
+        realm_id: Option<RealmId>,
+        user_id: UserId,
+    ) -> Result<(), String> {
+        if self.user_id != user_id {
+            return Err("Owner does not match".into());
+        }
+
+        if let Some(new_credits) = credits {
+            self.credits = new_credits;
+        }
+
+        if let Some(id) = realm_id {
+            if id.is_empty() {
+                self.realm_id = None;
+            } else {
+                self.realm_id = Some(id);
+            }
+        }
```

Comment:
I think it's nice to have. QR codes will be in phisical locations, so it's nice to have ability to change realm on them.

Reply:
When is such an invite expected to be clean up?

Reply:
It will never clean up. We have validation against creating invites over balance, so it won't be infinite.

 We could cleanup invites that run out of credits and were not used in long time?

Reply:
Ok, maybe it's not urgent right now.